File: go/mailpit_test.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/smtp"
	"strings"
	"testing"
	"time"
)

/**
 * Example of integration tests with Mailpit in Go.
 * Uses only the standard library:
 * - testing: test runner
 * - net/http: client for Mailpit API
 * - net/smtp: client for sending emails
 * - encoding/json: to process API responses
 */

const mailpitAPI = "http://localhost:8025/api/v1"

// Structs to map the Mailpit API response
type MailpitMessages struct {
	Total    int `json:"total"`
	Messages []struct {
		ID      string `json:"ID"`
		Subject string `json:"Subject"`
		From    struct {
			Address string `json:"Address"`
		} `json:"From"`
		To []struct {
			Address string `json:"Address"`
		} `json:"To"`
	} `json:"messages"`
}

type MailpitMessageDetail struct {
	ID   string `json:"ID"`
	HTML string `json:"HTML"`
	Text string `json:"Text"`
}

func TestMailpitIntegration(t *testing.T) {
	// 1. Clear existing emails
	t.Run("Initial Cleanup", func(t *testing.T) {
		req, _ := http.NewRequest(http.MethodDelete, mailpitAPI+"/messages", nil)
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatalf("Error cleaning messages: %v", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			t.Errorf("Expected status 200 on delete, got %d", resp.StatusCode)
		}
	})

	// 2. Send email via SMTP
	t.Run("Email Sending", func(t *testing.T) {
		from := "go-test@example.com"
		to := []string{"go-recipient@example.com"}
		subject := "Go API Test 🐹"
		body := "This email was sent during the Go integration test."

		msg := fmt.Sprintf("From: %s\r\nTo: %s\r\nSubject: %s\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n<b>%s</b>",
			from, to[0], subject, body)

		err := smtp.SendMail("localhost:1025", nil, from, to, []byte(msg))
		if err != nil {
			t.Fatalf("Error sending email via SMTP: %v", err)
		}
	})

	// Brief pause for Mailpit indexing
	time.Sleep(300 * time.Millisecond)

	var lastMessageID string

	// 3. Validate if the email was received via API
	t.Run("Validate Receipt", func(t *testing.T) {
		resp, err := http.Get(mailpitAPI + "/messages")
		if err != nil {
			t.Fatalf("Error fetching messages: %v", err)
		}
		defer resp.Body.Close()

		var result MailpitMessages
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			t.Fatalf("Error decoding JSON: %v", err)
		}

		if result.Total != 1 {
			t.Errorf("Expected 1 message, found %d", result.Total)
		}

		msg := result.Messages[0]
		if msg.Subject != "Go API Test 🐹" {
			t.Errorf("Incorrect subject: %s", msg.Subject)
		}

		if msg.To[0].Address != "go-recipient@example.com" {
			t.Errorf("Incorrect recipient: %s", msg.To[0].Address)
		}

		lastMessageID = msg.ID
		fmt.Printf("✅ Email found! ID: %s\n", lastMessageID)
	})

	// 4. Validate detailed content
	t.Run("Validate Content", func(t *testing.T) {
		if lastMessageID == "" {
			t.Skip("Skipping as we don't have the message ID")
		}

		resp, err := http.Get(mailpitAPI + "/message/" + lastMessageID)
		if err != nil {
			t.Fatalf("Error fetching message details: %v", err)
		}
		defer resp.Body.Close()

		var detail MailpitMessageDetail
		if err := json.NewDecoder(resp.Body).Decode(&detail); err != nil {
			t.Fatalf("Error decoding details: %v", err)
		}

		if !strings.Contains(detail.HTML, "sent during the Go integration test") {
			t.Errorf("HTML content does not contain the expected text")
		}
		fmt.Println("✅ HTML content validated successfully.")
	})
}

File: node/mailpit.test.js
const test = require('node:test');
const assert = require('node:assert');
const nodemailer = require('nodemailer');

/**
 * Example tests using the Mailpit API.
 * This file demonstrates how to run email integration tests without external 
 * testing libraries (besides nodemailer already installed for sending).
 * 
 * We use:
 * - node:test (Node.js native test runner)
 * - fetch (Node.js 18+ native API)
 */

const MAILPIT_API = 'http://localhost:8025/api/v1';

// SMTP transport configuration for Mailpit
const transporter = nodemailer.createTransport({
  host: 'localhost',
  port: 1025,
  secure: false,
  auth: null
});

test('Mailpit API - Integration Test Demonstration', async (t) => {

  await t.test('1. Initial cleanup: should delete all messages', async () => {
    console.log('🧹 Cleaning inbox...');
    const response = await fetch(`${MAILPIT_API}/messages`, { method: 'DELETE' });
    assert.strictEqual(response.status, 200, 'API should return 200 on delete');
    
    const check = await fetch(`${MAILPIT_API}/messages`);
    const data = await check.json();
    assert.strictEqual(data.total, 0, 'Inbox should be empty');
  });

  await t.test('2. Sending: should send an email via SMTP', async () => {
    console.log('📧 Sending test email...');
    const info = await transporter.sendMail({
      from: 'api-test@example.com',
      to: 'you@example.com',
      subject: 'API Test Subject 🚀',
      text: 'This is an email sent via Node.js to be validated by the Mailpit API.',
      html: '<b>HTML content tested!</b>'
    });

    assert.ok(info.messageId, 'Should have a messageId after sending');
  });

  await t.test('3. Validation: should find the email via Mailpit API', async () => {
    // Wait a brief moment for Mailpit to process the message
    await new Promise(r => setTimeout(r, 300));

    console.log('🔍 Searching for email in API...');
    const response = await fetch(`${MAILPIT_API}/messages`);
    const data = await response.json();

    assert.strictEqual(data.total, 1, 'Should have exactly 1 message in the inbox');
    const msg = data.messages[0];

    assert.strictEqual(msg.Subject, 'API Test Subject 🚀');
    assert.strictEqual(msg.To[0].Address, 'you@example.com');
    assert.strictEqual(msg.From.Address, 'api-test@example.com');
    
    console.log(`✅ Email found! Mailpit ID: ${msg.ID}`);
  });

  await t.test('4. Details: should retrieve and validate HTML content', async () => {
    // First we get the list to obtain the current message ID
    const listRes = await fetch(`${MAILPIT_API}/messages`);
    const listData = await listRes.json();
    const messageId = listData.messages[0].ID;

    // We fetch the full message details
    const msgRes = await fetch(`${MAILPIT_API}/message/${messageId}`);
    const msgData = await msgRes.json();

    assert.ok(msgData.HTML.includes('HTML content tested!'), 'HTML content should contain the expected string');
    console.log('✅ HTML content validated successfully.');
    console.log('---');
  });

});
